data/tutorials/language/0lg_08_basic_datatypes.md

```diff
@@ -0,0 +1,689 @@
+# fst (3, 'a');;
+- : int = 3
+
+# snd (3, 'a');;
+- : char = 'a'
+```
+
+In the standard library, both are defined using pattern matching. Here is how a function extracts the third element of the product of four types:
+```ocaml
+# let f x = match x with (a, b, c, d) -> c;;
+val f : 'a * 'b * 'c * 'd -> 'c = <fun>
+```
+
+Note that the product type operator `*` is not associative. Types `int * char * bool`, `int * (char * bool)`, and `(int * char) * bool` are not the same. The values `(42, 'a', true)`, `(42, ('a', true))`, and `((42, 'a'), true)` are not equal.
+
+### Functions
+
+The type of functions from type `a` to type `b` is written `a -> b`. Here are a few examples:
+```ocaml
+# fun x -> x * x;;
+- : int -> int = <fun>
+
+# (fun x -> x * x) 9;;
+- : int = 81
+```
+
+The first expression is an anonymous function of type `int -> int`. The type is inferred from the expression `x * x`, which must be of type `int` since `*` is an operator that returns an `int`. The `<fun>` printed in place of the value is a token, meaning functions don't have a value to be displayed. This is because if they have been compiled their code is no longer available.
+
+The second expression is function application. The parameter `9` is applied, and the result `81`` is returned.
+
+```ocaml
+# fun x -> x;;
+- : 'a -> 'a = <fun>
+
+# (fun x -> x) 42;;
+- : int = 42
+
+# (fun x -> x) "This is really disco!";;
+- : string = "This is really disco!"
+```
+
+The first expression is another anonymous function. It is the identity function, and it returns its argument unchanged. This function can be applied to anything, and anything can be returned unchanged. This means that function's parameter can be of any type, and the result must have the same type. The same code can be applied to data of different types. This is called _polymorphism_.
+
+This is what is indicated by the `'a` in the type (pronounced as the Greek letter α, “alpha”). This is a _type variable_. It means values of any type can be passed to the function. When that happens, their type is substituted for the type variable. This also expresses identity has the same input and output type, whatever it may be.
+
+The two following expressions show that the identity function can indeed be applied to parameters of different types.
+
+```ocaml
+# let f = fun x -> x * x;;
+f : int -> int = <fun>
+
+# f 9;;
+- : int = 81
+```
+
+Defining a function is the same as giving a name to any value. This is illustrated in the first expression:
+
+```ocaml
+# let g x = x * x;;
+g : int -> int = <fun>
+
+# g 9;;
+- : int = 81
+```
+
+When writing in OCaml, a lot of functions are written. The function `g` is defined here using a shorter, more common syntax, and maybe a more intuitive syntax.
+
+In OCaml, functions may terminate without returning a value of the expected type by throwing an exception, which does not appear in its type. There is no way to know if a function may raise an exception without inspecting its code.
```

Comment:
Should this tutorial briefly explain when it's appropriate to use exceptions and not `Result`? As a beginner reading this tutorial, I would be wondering: given that exceptions are not expressed in the type (which seems like a downside), when is it still worth using them?

Reply:
TODO: Let's link to the error handling tutorial, which address that question